compiler/rustc_trait_selection/src/traits/mod.rs

```diff
@@ -759,48 +760,38 @@ fn vtable_trait_first_method_offset<'tcx>(
 pub fn vtable_trait_upcasting_coercion_new_vptr_slot(
     tcx: TyCtxt<'tcx>,
     key: (
-        ty::PolyTraitRef<'tcx>, // trait owning vtable
-        ty::PolyTraitRef<'tcx>, // super trait ref
+        Ty<'tcx>, // trait object type whose trait owning vtable
+        Ty<'tcx>, // trait object for supertrait
     ),
 ) -> Option<usize> {
-    let (trait_owning_vtable, super_trait_ref) = key;
-    let super_trait_did = super_trait_ref.def_id();
-    // FIXME: take substsref part into account here after upcasting coercion allows the same def_id occur
-    // multiple times.
+    let (source, target) = key;
+    debug_assert!(matches!(&source.kind(), &ty::Dynamic(..)));
+    debug_assert!(matches!(&target.kind(), &ty::Dynamic(..)));
```

Comment:
can we add a `assert!(!source.needs_infer())` and same for `target`

Reply:
Sure, done!